File: DFS/#617 Merge Two Binary Trees.cpp
//No creation of new tre -> track t1

class Solution {
public:
    TreeNode* mergeTrees(TreeNode* t1, TreeNode* t2) {
        if(!t1 && !t2) return NULL;
        if(!t1 || !t2) {
            return t1 ? t1 : t2;
        }
        t1->val += t2->val;
        t1->left = mergeTrees(t1->left,t2->left);
        t1->right = mergeTrees(t1->right,t2->right);
        return t1;
        
    }
    
    
};
// Create new tree =
class Solution {
public:
    TreeNode* mergeTrees(TreeNode* t1, TreeNode* t2) {
        if(!t1&&!t2) return NULL;
        if(!t1 && t2) return t2;
        if(t1 && !t2) return t1;
        TreeNode* node = new TreeNode();
        node->val = t1->val + t2->val;
        node->left = mergeTrees(t1->left,t2->left);
        node->right = mergeTrees(t1->right,t2->right);
        return node;
        
    }
    
    
};

File: Tree/#617. Merge Two Binary Trees.py
class Solution:
    def mergeTrees(self, root1: Optional[TreeNode], root2: Optional[TreeNode]) -> Optional[TreeNode]:
        return self.dfs(root1, root2)

    def dfs(self, r1, r2):
        if r1 and r2:
            r1.val += r2.val
        elif r1:
            return r1
        elif r2:
            return r2
        else:
            return None
        
        r1.left = self.dfs(r1.left, r2.left)
        r1.right = self.dfs(r1.right, r2.right)
        return r1
        